File: infra/components/registry.py
import pulumi
from pulumi import Inputs, ResourceOptions
from pulumi_azure_native import containerregistry

class RegistryWithCredentials(pulumi.ComponentResource):
    def __init__(
        self,
        name: str,
        resource_group_name: pulumi.Input[str],
        props: Inputs | None = None,
        opts: ResourceOptions | None = None,
        remote: bool = False
    ) -> None:
        super().__init__('RegistryWithCredentials', name, props, opts, remote)

        # Create a container registry.
        self.registry = containerregistry.Registry(
            f"{name}Registry",
            containerregistry.RegistryArgs(
                resource_group_name=resource_group_name,
                admin_user_enabled=True,
                sku=containerregistry.SkuArgs(
                    name=containerregistry.SkuName.BASIC,
                ),
            ),
        )

        # Fetch login credentials for the registry.
        self.credentials = containerregistry.list_registry_credentials_output(
            resource_group_name=resource_group_name,
            registry_name=self.registry.name,
        )

        self.login_server = self.registry.login_server

    @property
    def username(self):
        return self.credentials.apply(lambda creds: creds.username)

    @property
    def password(self):
        return self.credentials.apply(lambda creds: creds.passwords[0].value)

File: infra/src/components/registry.ts
import * as pulumi from "@pulumi/pulumi";
import * as azure from "@pulumi/azure-native";

// TODO Remove/Use this component
type RegistryArgs = {
	resourceGroupName: string;
}

export class Registry extends pulumi.ComponentResource {
	private resourceGroupName: string;
	registry: azure.containerregistry.Registry

	constructor(name: string, args: RegistryArgs, opts?: pulumi.ComponentResourceOptions) {
		super("infra_shared_components:components:Registry", name, {}, opts);

		this.registry = new azure.containerregistry.Registry(
			`${name}Registry`,
			{
				resourceGroupName: args.resourceGroupName,
				adminUserEnabled: true,
				sku: {
					name: azure.containerregistry.SkuName.Basic,
				}
			}
		);

		this.resourceGroupName = args.resourceGroupName;

		this.registerOutputs({});
	}

	async getCredentials(): Promise<{ username: pulumi.Output<string>, password: pulumi.Output<string>}> {
		const creds = this.registry.name.apply((registryName) => {
			return azure.containerregistry.listRegistryCredentials({
				registryName,
				resourceGroupName: this.resourceGroupName,
			})
		})

		return {
			username: creds.username!.apply(u => u!),
			password: creds.passwords!.apply(p => p![0].value!),
		}
	}
}
